pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/invertedindex/TextIndexHandler.java

```diff
@@ -174,18 +175,46 @@ private void createTextIndexForColumn(SegmentDirectory.Writer segmentWriter, Col
     // segmentDirectory is indicated to us by SegmentDirectoryPaths, we create lucene index there. There is no
     // further need to move around the lucene index directory since it is created with correct directory structure
     // based on segmentVersion.
-    try (ForwardIndexReader forwardIndexReader = LoaderUtils.getForwardIndexReader(segmentWriter, columnMetadata);
-        ForwardIndexReaderContext readerContext = forwardIndexReader.createContext();
-        TextIndexCreator textIndexCreator = indexCreatorProvider.newTextIndexCreator(IndexCreationContext.builder()
-            .withColumnMetadata(columnMetadata).withIndexDir(segmentDirectory).build().forTextIndex(_fstType, true))) {
-      if (columnMetadata.isSingleValue()) {
-        processSVField(segmentWriter, hasDictionary, forwardIndexReader, readerContext, textIndexCreator, numDocs,
-            columnMetadata);
+    try (TextIndexCreator textIndexCreator = indexCreatorProvider.newTextIndexCreator(
+        IndexCreationContext.builder().withColumnMetadata(columnMetadata).withIndexDir(segmentDirectory).build()
+            .forTextIndex(_fstType, true))) {
+      boolean forwardIndexDisabled = !segmentWriter.hasIndexFor(columnName, ColumnIndexType.FORWARD_INDEX);
+      if (forwardIndexDisabled) {
+        try (Dictionary dictionary = LoaderUtils.getDictionary(segmentWriter, columnMetadata)) {
+          // Create the text index if the dictionary length is 1 as this is for a default column (i.e. newly added
+          // column). For existing columns it is not possible to create the text index without forward index
+          Preconditions.checkState(dictionary.length() == 1, String.format("Creating text index for forward index "
```

Comment:
Suggest not adding this part but directly throwing exception. I don't see the point of special handling generating text index for a column with all same values. Same for other index handlers.

Reply:
Hi @Jackie-Jiang what do you mean by directly throwing exception?

Without adding this special handling, creation of these indexes for default columns will fail. Today when a new column is added, a default forward index is created. On the reload path I'm skipping the creation of the forward index when this feature is disabled for such default columns. I still create a dictionary though. I do want to create other indices since queries may fail if such segments are accessed and the expected indexes don't exist, right? Or am I missing something here? (i did see the segment preprocessor tests fail when I didn't do special handling for the inverted index for example since the default handlers directly access the forward index to create the auxiliary index)

Reply:
Hmm.. That is unexpected. When the dictionary size is 1, the column should be sorted, and we shouldn't skip the forward index

Reply:
So we actually evaluated two approaches for this part:

- Create the forward index, allow it to be marked as sorted. This will skip creation of the inverted index, and all other index handlers will kick in and create the index off of the forward index we create. (some index handlers rely on the forward index to construct the index)
    - Our main concern with this approach was that the segments will be inconsistent, some will have a forward index and others won't. This may give odd behavior at query time depending on the segment touched (e.g. some queries go fine with a given filter but don't go through with a slightly different filter). We wanted to avoid this.
    - We had also explored an option to identify such queries on the Broker or Server planning stage but quickly found that this became very ugly and had lots of exceptions and was error prone.
- Don't create the forward index but do create the dictionary (just like the normal segment creation path). Modify the index handlers to create the index for such a default column if forward index is disabled without relying on the the forward index. In such cases we expect exactly 1 value anyways. This will give a consistent experience to users who run queries.
    - This has the disadvantage of doing some special handling in the index handlers for this scenario since the forward index won't be available anymore.

Hope the above explains our thinking about the approaches and why we decided to go with the second one. Let me know if you'd like to discuss this in more detail. cc @siddharthteotia

Reply:
I don't follow the second approach. For the default column where all the values are the same for a column, we should create sorted index and a single entry dictionary. This is the same behavior as generating a new segment with a column of all the same values.
IMO second approach will actually cause inconsistent behavior because it will have inverted index with only a single entry, and the segment generated during the reload or from the raw data without a column will be different.
I strongly suggest generating sorted index + dictionary for the default column.

Reply:
For the MV default column, since we don't generate sorted index for it, we can consider directly creating inverted index in the default column handler.
Essentially we should try to keep the behavior the same for reloaded segment and segment generated without the data for the default column. That way we don't need any special handling for other index handlers, and the logic will be cleaner.

Reply:
@Jackie-Jiang and @siddharthteotia and I discussed these remaining comments and decided to do the following:

- For SV columns the default column should always create a sorted forward index for consistency with the normal segment creation path. We agreed that the query inconsistency that users might see is acceptable since queries may work when they are expected not to work and not the other way around.
- Default multi-value columns should not have a forward index in the long run. For simplifying this PR we decided to ignore the `forwardIndexDisabled` flag for the most part on the reload code path.
- The next change we need to make is address reload code path in general within which we'll add handling for the default column on reload path as well. This will ensure we can use a consistent mechanism to handle the MV default column case above without resorting to special handling in the IndexHandlers.